<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Цикл-for" data-toc-modified-id="Цикл-for-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Цикл for</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-1.1"><span class="toc-item-num">1.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-1.2"><span class="toc-item-num">1.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-1.3"><span class="toc-item-num">1.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-1.4"><span class="toc-item-num">1.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-1.5"><span class="toc-item-num">1.5&nbsp;&nbsp;</span>5</a></span></li><li><span><a href="#6" data-toc-modified-id="6-1.6"><span class="toc-item-num">1.6&nbsp;&nbsp;</span>6</a></span></li><li><span><a href="#7" data-toc-modified-id="7-1.7"><span class="toc-item-num">1.7&nbsp;&nbsp;</span>7</a></span></li><li><span><a href="#8" data-toc-modified-id="8-1.8"><span class="toc-item-num">1.8&nbsp;&nbsp;</span>8</a></span></li></ul></li><li><span><a href="#Цикл-for:-функция-range" data-toc-modified-id="Цикл-for:-функция-range-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Цикл for: функция range</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-2.1"><span class="toc-item-num">2.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-2.2"><span class="toc-item-num">2.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-2.3"><span class="toc-item-num">2.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-2.4"><span class="toc-item-num">2.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-2.5"><span class="toc-item-num">2.5&nbsp;&nbsp;</span>5</a></span></li></ul></li><li><span><a href="#Частые-сценарии" data-toc-modified-id="Частые-сценарии-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Частые сценарии</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-3.1"><span class="toc-item-num">3.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-3.2"><span class="toc-item-num">3.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-3.3"><span class="toc-item-num">3.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-3.4"><span class="toc-item-num">3.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-3.5"><span class="toc-item-num">3.5&nbsp;&nbsp;</span>5</a></span></li><li><span><a href="#6" data-toc-modified-id="6-3.6"><span class="toc-item-num">3.6&nbsp;&nbsp;</span>6</a></span></li><li><span><a href="#7" data-toc-modified-id="7-3.7"><span class="toc-item-num">3.7&nbsp;&nbsp;</span>7</a></span></li><li><span><a href="#8" data-toc-modified-id="8-3.8"><span class="toc-item-num">3.8&nbsp;&nbsp;</span>8</a></span></li><li><span><a href="#9" data-toc-modified-id="9-3.9"><span class="toc-item-num">3.9&nbsp;&nbsp;</span>9</a></span></li><li><span><a href="#10" data-toc-modified-id="10-3.10"><span class="toc-item-num">3.10&nbsp;&nbsp;</span>10</a></span></li><li><span><a href="#11" data-toc-modified-id="11-3.11"><span class="toc-item-num">3.11&nbsp;&nbsp;</span>11</a></span></li></ul></li><li><span><a href="#Цикл-while" data-toc-modified-id="Цикл-while-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Цикл while</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>5</a></span></li><li><span><a href="#6" data-toc-modified-id="6-4.6"><span class="toc-item-num">4.6&nbsp;&nbsp;</span>6</a></span></li><li><span><a href="#7" data-toc-modified-id="7-4.7"><span class="toc-item-num">4.7&nbsp;&nbsp;</span>7</a></span></li></ul></li><li><span><a href="#Цикл-while:-обработка-цифр-числа" data-toc-modified-id="Цикл-while:-обработка-цифр-числа-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Цикл while: обработка цифр числа</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-5.1"><span class="toc-item-num">5.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-5.2"><span class="toc-item-num">5.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-5.3"><span class="toc-item-num">5.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-5.4"><span class="toc-item-num">5.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-5.5"><span class="toc-item-num">5.5&nbsp;&nbsp;</span>5</a></span></li><li><span><a href="#6" data-toc-modified-id="6-5.6"><span class="toc-item-num">5.6&nbsp;&nbsp;</span>6</a></span></li><li><span><a href="#7" data-toc-modified-id="7-5.7"><span class="toc-item-num">5.7&nbsp;&nbsp;</span>7</a></span></li></ul></li><li><span><a href="#break,-continue-и-else" data-toc-modified-id="break,-continue-и-else-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>break, continue и else</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-6.1"><span class="toc-item-num">6.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-6.2"><span class="toc-item-num">6.2&nbsp;&nbsp;</span>2</a></span></li></ul></li><li><span><a href="#Поиск-ошибок-и-ревью-кода" data-toc-modified-id="Поиск-ошибок-и-ревью-кода-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Поиск ошибок и ревью кода</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-7.1"><span class="toc-item-num">7.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-7.2"><span class="toc-item-num">7.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-7.3"><span class="toc-item-num">7.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-7.4"><span class="toc-item-num">7.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-7.5"><span class="toc-item-num">7.5&nbsp;&nbsp;</span>5</a></span></li><li><span><a href="#6" data-toc-modified-id="6-7.6"><span class="toc-item-num">7.6&nbsp;&nbsp;</span>6</a></span></li></ul></li><li><span><a href="#Вложенные-циклы.-Часть-1" data-toc-modified-id="Вложенные-циклы.-Часть-1-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Вложенные циклы. Часть 1</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-8.1"><span class="toc-item-num">8.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-8.2"><span class="toc-item-num">8.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-8.3"><span class="toc-item-num">8.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-8.4"><span class="toc-item-num">8.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-8.5"><span class="toc-item-num">8.5&nbsp;&nbsp;</span>5</a></span></li><li><span><a href="#6" data-toc-modified-id="6-8.6"><span class="toc-item-num">8.6&nbsp;&nbsp;</span>6</a></span></li><li><span><a href="#7" data-toc-modified-id="7-8.7"><span class="toc-item-num">8.7&nbsp;&nbsp;</span>7</a></span></li><li><span><a href="#8" data-toc-modified-id="8-8.8"><span class="toc-item-num">8.8&nbsp;&nbsp;</span>8</a></span></li></ul></li><li><span><a href="#Вложенные-циклы.-Часть-2" data-toc-modified-id="Вложенные-циклы.-Часть-2-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Вложенные циклы. Часть 2</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-9.1"><span class="toc-item-num">9.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-9.2"><span class="toc-item-num">9.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-9.3"><span class="toc-item-num">9.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-9.4"><span class="toc-item-num">9.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-9.5"><span class="toc-item-num">9.5&nbsp;&nbsp;</span>5</a></span></li><li><span><a href="#6" data-toc-modified-id="6-9.6"><span class="toc-item-num">9.6&nbsp;&nbsp;</span>6</a></span></li><li><span><a href="#7" data-toc-modified-id="7-9.7"><span class="toc-item-num">9.7&nbsp;&nbsp;</span>7</a></span></li></ul></li><li><span><a href="#Итоговая-работа-на-циклы" data-toc-modified-id="Итоговая-работа-на-циклы-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Итоговая работа на циклы</a></span><ul class="toc-item"><li><span><a href="#1" data-toc-modified-id="1-10.1"><span class="toc-item-num">10.1&nbsp;&nbsp;</span>1</a></span></li><li><span><a href="#2" data-toc-modified-id="2-10.2"><span class="toc-item-num">10.2&nbsp;&nbsp;</span>2</a></span></li><li><span><a href="#3" data-toc-modified-id="3-10.3"><span class="toc-item-num">10.3&nbsp;&nbsp;</span>3</a></span></li><li><span><a href="#4" data-toc-modified-id="4-10.4"><span class="toc-item-num">10.4&nbsp;&nbsp;</span>4</a></span></li><li><span><a href="#5" data-toc-modified-id="5-10.5"><span class="toc-item-num">10.5&nbsp;&nbsp;</span>5</a></span></li><li><span><a href="#6" data-toc-modified-id="6-10.6"><span class="toc-item-num">10.6&nbsp;&nbsp;</span>6</a></span></li><li><span><a href="#7" data-toc-modified-id="7-10.7"><span class="toc-item-num">10.7&nbsp;&nbsp;</span>7</a></span></li></ul></li></ul></div>

# Циклы for и while

## Цикл for

### 1

Напишите программу, которая выводит слова «Python is awesome!» (без кавычек) 10 раз.

<img src="https://ucarecdn.com/a641f6da-809b-4895-8b1d-3db32a0e8f26/" width="400" height="400">

In [67]:
for i in range(10):
    print("Python is awesome!")

Python is awesome!
Python is awesome!
Python is awesome!
Python is awesome!
Python is awesome!
Python is awesome!
Python is awesome!
Python is awesome!
Python is awesome!
Python is awesome!


### 2

Дано предложение и количество раз которое его надо повторить. Напишите программу, которая повторяет данное предложение нужное количество раз.

In [6]:
text = input()
number = int(input())
for i in range(number):
    print(text)

Век живи - век учись.
3
Век живи - век учись.
Век живи - век учись.
Век живи - век учись.


### 3

Напишите программу, которая использует ровно три цикла for для печати следующей последовательности символов:

AAA
AAA
AAA
AAA
AAA
AAA
BBBB
BBBB
BBBB
BBBB
BBBB
E
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
G

In [68]:
for i in range(6):
    print('AAA')
for t in range(5):
    print('BBBB')
print('E')
for y in range(9):
    print('TTTTT')
print('G')

AAA
AAA
AAA
AAA
AAA
AAA
BBBB
BBBB
BBBB
BBBB
BBBB
E
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
TTTTT
G


### 4

На вход программе подается натуральное число $n$. Напишите программу, которая печатает звездный прямоугольник размерами $n×19$.

На вход программе подаётся натуральное число $n∈[1;20]$ — высота звездного прямоугольника.

In [7]:
n = int(input())
for star in range(n):
    print('*' * 19)

3
*******************
*******************
*******************


### 5

Напишите программу, которая считывает одну строку текста и выводит 10 строк, пронумерованных от 0 до 9, каждая с указанной строкой текста.

In [8]:
text = input()
for i in range(10):
    print(i, text)

мой текст
0 мой текст
1 мой текст
2 мой текст
3 мой текст
4 мой текст
5 мой текст
6 мой текст
7 мой текст
8 мой текст
9 мой текст


### 6

На вход программе подается натуральное число $n$. Напишите программу, которая для каждого из чисел от $0$ до $n$ (включительно) выводит фразу: «Квадрат числа [число] равен [число]» (без кавычек).

In [9]:
n = int(input())
for number in range(n + 1):
    print(f'Квадрат числа {number} равен {number ** 2}')

5
Квадрат числа 0 равен 0
Квадрат числа 1 равен 1
Квадрат числа 2 равен 4
Квадрат числа 3 равен 9
Квадрат числа 4 равен 16
Квадрат числа 5 равен 25


### 7

На вход программе подается натуральное число $n(n≥2)$ – катет прямоугольного равнобедренного треугольника.

Напишите программу, которая выводит звездный треугольник в соответствии с примером.

In [11]:
num = int(input())
for k in range(num):
    print('*' * (num - k))

3
***
**
*


### 8

На вход программе подается три натуральных числа $m$, $p$, $n$:

* $m$: стартовое количество организмов;
* $p$: среднесуточное увеличение в %;
* $n$: количество дней для размножения.

Напишите программу, которая предсказывает размер популяции организмов. Программа должна выводить размер популяции в каждый день, начиная с $1$ и заканчивая n-м днем.

In [17]:
m, p, n = int(input()), int(input()), int(input())
for day in range(1, n + 1):
    print(day, m)
    m = m + m * p / 100

120
25
4
1 120
2 150.0
3 187.5
4 234.375


## Цикл for: функция range

### 1

Даны два целых числа $m$ и $n$ $(m≤n)$. Напишите программу, которая выводит все числа от $m$ до $n$ включительно.

In [69]:
m, n = int(input()), int(input())
for i in range (m, n+1):
    print(i)

1
6
1
2
3
4
5
6


### 2

Даны два целых числа $m$ и $n$. Напишите программу, которая выводит все числа от $m$ до $n$ включительно в порядке возрастания, если $m < n$, или в порядке убывания в противном случае.

In [23]:
m, n = int(input()), int(input())
if m > n:
    for i in range(m, n-1, -1):
        print(i)
elif m < n:
    for g in range(m, n+1):
        print(g)
elif m == n:
    print(m)

7
1
7
6
5
4
3
2
1


### 3

Даны два целых числа $m$ и $n$ $(m > n)$. Напишите программу, которая выводит все нечетные числа от $m$ до $n$ включительно в порядке убывания.

In [ ]:
m, n = int(input()), int(input())
if m % 2 == 0:
    for i in range(m-1, n-1, -2):
        print(i)
else:
    for k in range(m, n-1, -2):
        print(k)

In [24]:
m, n = int(input()), int(input())
for i in range(m - 1 + m % 2, n - 1, -2):
    print(i)

5
2
5
3


### 4

Даны два натуральных числа $m$ и $n (m≤n)$. Напишите программу, которая выводит все числа от $m$ до $n$ включительно удовлетворяющие хотя бы одному из условий:

* число кратно 17;
* число оканчивается на 9;
* число кратно 3 и 5 одновременно.

In [25]:
m, n = int(input()), int(input())
for i in range(m, n+1):
    if i % 17 == 0 or i % 10 == 9 or i % 15 == 0:
        print(i)

1
20
9
15
17
19


### 5

Дано натуральное число $n$. Напишите программу, которая выводит таблицу умножения на $n$.

In [6]:
n = int(input())
for i in range(1, 11):
    print(f'{n} x {i} =', n * i)

5
5 x 1 = 5
5 x 2 = 10
5 x 3 = 15
5 x 4 = 20
5 x 5 = 25
5 x 6 = 30
5 x 7 = 35
5 x 8 = 40
5 x 9 = 45
5 x 10 = 50


## Частые сценарии

**Подсчет количества**

In [ ]:
counter1 = 0 # переменная-счетчик 1
counter2 = 0 # переменная-счетчик 2
for i in range(10):
    num = int(input())
    if num > 10:
        counter1 = counter1 + 1
    if num == 0:
        counter2 = counter2 + 1
print('Было введено', counter1, 'чисел, больших 10.')
print('Было введено', counter2, 'нулей.' )

In [ ]:
counter = 0 # переменная-счетчик чисел из диапазона [1; 100], квадрат которых оканчивается на 4
for i in range(1, 101):
    if i**2 % 10 == 4:
        counter = counter + 1
print(counter)

**Вычисление суммы и произведения**

In [ ]:
total = 0 # считывает 10 чисел и определяет сумму тех из них, которые больше 10
for i in range(10):
    num = int(input())
    if num > 10:
        total = total + num
print('Сумма чисел больших 10 равна',  total)

**Обмен значений переменных**

In [ ]:
temp = x
x = y
y = temp

In [ ]:
x, y = y, x # лучший вариант

**Сигнальные метки**

In [ ]:
num = int(input())
flag = True

for i in range(2, num):
    if num % i == 0:        #  если исходное число делится на какое-либо отличное от 1 и самого себя
        flag = False

if num == 1:
    print('Это единица, она не простая и не составная') 
elif flag == True:
    print('Число простое')
else:
    print('Число составное')

**Максимум и минимум**

In [ ]:
largest = -1
for i in range(10):
    num = int(input())    
    if num > largest:
        largest = num
print('Наибольшее число равно', largest) 

In [ ]:
largest = int(input())  # принимаем первое число за максимальное
for i in range(9):
    num = int(input())
    if num > largest:
        largest = num
print('Наибольшее число равно', largest) 

### 1

На вход программе подаются два целых числа $a$ и $b$ $(a≤b)$. Напишите программу, которая подсчитывает количество чисел в диапазоне от $a$ до $b$ включительно, куб которых оканчивается на $4$ или $9$.

In [10]:
a, b = int(input()), int(input())
counter = 0
for i in range(a, b + 1):
    if i**3 % 10 == 4 or i**3 % 10 == 9:
        counter = counter + 1
print(counter)

1
4
1


### 2

На вход программе подается натуральное число $n$, а затем $n$ целых чисел, каждое на отдельной строке. Напишите программу, которая подсчитывает сумму введенных чисел. 

In [11]:
n = int(input())
sum = 0
for i in range(n):
    num = int(input())
    sum += num
print(sum)

3
0
5
2
7


### 3

На вход программе подается натуральное число $n$. Напишите программу, которая вычисляет значение выражения
$\left(1+\dfrac12 + \dfrac13 + \ldots + \dfrac{1}{n} \right) - \ln (n)$.

Для вычисления натурального логарифма воспользуйтесь функцией ```log(n)```, которая находится в модуле ```math```.

In [13]:
import math
n = int(input())
sum = 0
for i in range(1, n + 1):
    num = 1 / i
    sum += num
print(sum - math.log(n))

4
0.6970389722134425


### 4

На вход программе подается натуральное число $n$. Напишите программу, которая подсчитывает сумму тех чисел от $1$ до $n$ (включительно), квадрат которых оканчивается на $2$, $5$ или $8$.

In [14]:
n = int(input())
sum = 0
for i in range(1, n+1):
    if i**2 % 10 == 2 or i**2 % 10 == 5 or i**2 % 10 == 8:
        sum += i
print(sum)

3
0


### 5

На вход программе подается натуральное число $n$. Напишите программу, которая вычисляет $n!$.

In [15]:
n = int(input())
factorial = 1
for i in range(1, n+1):
    factorial *= i
print(factorial)

4
24


### 6

Напишите программу, которая считывает 10 чисел и выводит произведение отличных от нуля чисел.

In [18]:
total = 1
for i in range(10):
    num = int(input())
    if num != 0:
        total *= num
print(total)

1
43
2
234
78
0
1
1
23
4
144409824


### 7

На вход программе подается натуральное число $n$. Напишите программу, которая вычисляет сумму всех его делителей.

In [19]:
sum = 0
n = int(input())
for i in range(1, n+1):
    if n % i == 0:
        sum += i
print(sum)

10
18


### 8

На вход программе подается натуральное число $n$. Напишите программу вычисления знакочередующей суммы 
$1-2+3-4+5-6 + \ldots + (-1)^{n+1}n$.

In [20]:
sum = 0
n = int(input())
for i in range(1, n+1):
    if i % 2 != 0:
        sum += i
    else:
        sum -= i
print(sum)

5
3


### 9

На вход программе подается натуральное число $n$, а затем $n$ различных натуральных чисел, каждое на отдельной строке. Напишите программу, которая выводит наибольшее и второе наибольшее число последовательности.

In [27]:
n = int(input())
largest_1st = 0
largest_2nd = 0
for i in range(n):
    num = int(input())
    if num > largest_1st:
        largest_2nd, largest_1st = largest_1st, num
    elif largest_2nd < num < largest_1st:
        largest_2nd = num
print(largest_1st, largest_2nd, sep='\n')

5
1
2
3
4
5
5
4


### 10

Напишите программу, которая считывает последовательность из 10 целых чисел и определяет, является ли каждое из них четным или нет.

In [29]:
flag = True
for i in range(10):
    num = int(input())
    if num % 2 != 0:
        flag = False
if flag == True:
    print('YES')
else:
    print('NO')

1
5
343
22
55
3
5
6
2
3
NO


### 11

Напишите программу, которая считывает натуральное число $n$ и выводит первые $n$ чисел последовательности Фибоначчи.

In [36]:
n = int(input())
num = 0
num_next = 1
for i in range(1, n+1):
    num, num_next = num_next, num+num_next
    print(num, end=' ')

5
1 1 2 3 5 

## Цикл while

### 1

На вход программе подается последовательность слов, каждое слово на отдельной строке. Концом последовательности является слово «КОНЕЦ» (без кавычек). Напишите программу, которая выводит члены данной последовательности.

In [41]:
word = str(input())
while word != 'КОНЕЦ':
    print(word)
    word = str(input())

рождение
рождение
жизнь
жизнь
КОНЕЦ


### 2

На вход программе подается последовательность слов, каждое слово на отдельной строке. Концом последовательности является слово «КОНЕЦ» или «конец» (большими или маленькими буквами, без кавычек). Напишите программу, которая выводит члены данной последовательности.

In [6]:
word = str(input())
while word != 'КОНЕЦ' and word != 'конец':
    print(word)
    word = str(input())

а
а
а
а
ы
ы
конец


### 3

На вход программе подается последовательность слов, каждое слово на отдельной строке. Концом последовательности является одно из трех слов: «стоп», «хватит», «достаточно» (маленькими буквами, без кавычек). Напишите программу, которая выводит общее количество членов данной последовательности.

In [9]:
word = str(input())
sum = 0
while word != 'стоп' and word != 'хватит' and word != 'достаточно':
    sum += 1
    word = str(input())
print(sum)

в
ы
стоп
2


### 4

На вход программе подается последовательность целых чисел делящихся на $7$, каждое число на отдельной строке. Концом последовательности является любое число не делящееся на $7$. Напишите программу, которая выводит члены данной последовательности.

In [10]:
num = int(input())
while num % 7 == 0:
    print(num)
    num = int(input())

7
7
14
14
0
0
2


### 5

На вход программе подается последовательность целых чисел, каждое число на отдельной строке. Концом последовательности является любое отрицательное число. Напишите программу, которая выводит сумму всех членов данной последовательности.

In [13]:
num = int(input())
sum = 0
while num >= 0:
    sum += num
    num = int(input())
print(sum)

2
3
-4
5


### 6

На вход программе подается последовательность целых чисел от $1$ до $5$, характеризующее оценку ученика, каждое число на отдельной строке. Концом последовательности является любое отрицательное число, либо число большее $5$. Напишите программу, которая выводит количество пятерок.

In [2]:
num = int(input())
sum = 0
while 6 > num >= 0:
    if num == 5:
        sum += 1
    num = int(input())
print(sum)

2
5
4
5
-2
2


### 7

Всем известно, что ведьмак способен одолеть любых чудовищ, однако его услуги обойдутся недешево, к тому же ведьмак не принимает купюры, он принимает только чеканные монеты. В мире ведьмака существуют монеты с номиналами $1$, $5$, $10$, $25$.

Напишите программу, которая определяет, какое минимальное количество чеканных монет нужно заплатить ведьмаку.

In [3]:
price = int(input())
counter = 0
while price >= 25:
    price -= 25
    counter += 1
while price >= 10:
    price -= 10
    counter += 1
while price >= 5:
    price -= 5
    counter += 1
while price >= 1:
    price -= 1
    counter += 1
print(counter)

49
7


## Цикл while: обработка цифр числа

### 1

Дано натуральное число. Напишите программу, которая выводит его цифры в столбик в обратном порядке.

In [6]:
n = int(input())
while n != 0:  # пока в числе есть цифры
    last_digit = n % 10  # получить последнюю цифру
    print(last_digit)
    n = n // 10  # удалить последнюю цифру из числа

345
5
4
3


### 2

Дано натуральное число. Напишите программу, которая меняет порядок цифр числа на обратный.

In [4]:
n = int(input())
while n != 0:
    last_digit = n % 10  
    print(last_digit, end='')
    n = n // 10  

3546
6453

### 3

Дано натуральное число $n$, $(n≥10)$. Напишите программу, которая определяет его максимальную и минимальную цифры.

In [21]:
n = int(input())
min_digit = n % 10
max_digit = n % 10
while n != 0:
    last_digit = n % 10  
    if last_digit > max_digit:
        max_digit = last_digit
    if last_digit < min_digit:
        min_digit = last_digit
    n = n // 10 
print('Максимальная цифра равна', max_digit)
print('Минимальная цифра равна', min_digit)

8945
Максимальная цифра равна 9
Минимальная цифра равна 4


### 4

Дано натуральное число. Напишите программу, которая вычисляет:

* сумму его цифр;
* количество цифр в нем;
* произведение его цифр;
* среднее арифметическое его цифр;
* его первую цифру;
* сумму его первой и последней цифры.

In [7]:
n = int(input())
sum, num, total = 0, 0, 1
le = len(str(n))
digit1 = (n % 10**le) // 10**(le-1)
digitsum = digit1 + n % 10  
while n != 0:
    last_digit = n % 10  
    sum += last_digit
    num += 1
    total *= last_digit
    n = n // 10 
print(sum, num, total, sum/le, digit1, digitsum, sep='\n')

5678
26
4
1680
6.5
5
13


### 5

Дано натуральное число $n$, $(n>9)$. Напишите программу, которая определяет его вторую (с начала) цифру.

In [8]:
n = int(input())
while n > 9:
    last_digit = n % 10
    n = n // 10
print(last_digit)

45672
5


### 6

Дано натуральное число. Напишите программу, которая определяет, состоит ли указанное число из одинаковых цифр.

In [10]:
n = int(input())
last_digit = n % 10
while last_digit == n % 10:
    last_digit = n % 10
    n = n // 10
if n == 0:
    print('YES')
else:
    print('NO')

223
NO


### 7

Дано натуральное число. Напишите программу, которая определяет, является ли последовательность его цифр при просмотре справа налево упорядоченной по неубыванию.

In [12]:
n = int(input())
last_digit = 0
while last_digit <= n % 10:
    last_digit = n % 10
    n = n // 10
if n == 0:
    print('YES')
else:
    print('NO')

54321
YES


## break, continue и else

**Примеры**

In [ ]:
# Напишем программу, определяющую, что число является простым:

num = int(input())
flag = True

for i in range(2, num):
    if num % i == 0:        #  если исходное число делится на какое-либо отличное от 1 и самого себя
        flag = False
        break               # останавливаем цикл если встретили делитель числа        

if flag == True:
    print('Число простое')
else:
    print('Число составное')
    
# Оператор прерывания цикла break позволяет ускорять программы, так как мы избавляемся от лишних итераций.

In [ ]:
# Напишем, программу, которая определяет, содержит ли введенное пользователем число, цифру 7.

num = int(input())
number = num
flag = False
while num != 0:
    last_digit = num % 10
    if last_digit == 7:
        flag = True
        break        # прерываем цикл, так как число гарантированно содержит цифру 7
    num //= 10

if flag == True:
    print('Число', number, 'содержит цифру 7')
else:
    print('Число', number, 'не содержит цифру 7')

In [ ]:
for i in range(1, 101):
    if i == 7 or i == 17 or i == 29 or i == 78:
        continue  # переходим на следующую итерацию
    print(i)

### 1

На вход программе подается число $n>1$. Напишите программу, которая выводит его наименьший отличный от $1$ делитель.

In [79]:
n = int(input())
for i in range(2, n+1):
    if n % i == 0:
        print(i)
        break

15
3


### 2

На вход программе подается натуральное число $n$. Напишите программу, которая выводит числа от $1$ до $n$ включительно за исключением:

* чисел от $5$ до $9$ включительно;
* чисел от $17$ до $37$ включительно;
* чисел от $78$ до $87$ включительно.

In [81]:
n = int(input())
for i in range(1, n+1):
    if i in range(5, 10) or i in range(17, 38) or i in range(78, 88):
        continue
    else:
        print(i)

10
1
2
3
4
10


## Поиск ошибок и ревью кода

### 1

На обработку поступает последовательность из $10$ целых чисел. Известно, что вводимые числа по абсолютной величине не превышают $10^6$. Нужно написать программу, которая выводит на экран количество неотрицательных чисел последовательности и их произведение. Если неотрицательных чисел нет, требуется вывести на экран «NO». Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их ровно $4$). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

**Примечание 1.** Число $x$ не превышает по абсолютной величине $10^6$, если $−10^6 ≤ x ≤ 10^6$.

**Примечание 2.** При необходимости вы можете добавить необходимые строки кода.

In [ ]:
count = 0
p = 1
for i in range(0, 10):
    x = int(input())
    if x >= 0:
        p = p * x
        count = count + 1
if count > 0:
    print(count)
    print(p)
else:
    print('NO')

### 2

На обработку поступает последовательность из $10$ целых чисел. Известно, что вводимые числа по абсолютной величине не превышают $10^6$. Нужно написать программу, которая выводит на экран сумму всех отрицательных чисел последовательности и максимальное отрицательное число в последовательности. Если отрицательных чисел нет, требуется вывести на экран «NO». Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их ровно $5$). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

**Примечание 1.** Число $x$ не превышает по абсолютной величине $10^6$, если $−10^6 ≤ x ≤ 10^6$.

**Примечание 2.** При необходимости вы можете добавить необходимые строки кода.

In [ ]:
mx = -10**6 - 1
s = 0
for i in range(10):
    x = int(input())
    if x < 0:
        s += x
    if 0 > x > mx:
        mx = x
if s < 0:
    print(s)
    print(mx)
else:
    print('NO')

### 3

На обработку поступает последовательность из $7$ целых чисел. Известно, что вводимые числа по абсолютной величине не превышают $10^6$. Нужно написать программу, которая подсчитывает и выводит сумму всех чётных чисел последовательности или $0$, если чётных чисел в последовательности нет. Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их ровно $4$). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

**Примечание 1.** Число $x$ не превышает по абсолютной величине $10^6$, если $−10^6 ≤ x ≤ 10^6$.

**Примечание 2.** При необходимости вы можете добавить необходимые строки кода.

In [ ]:
s = 0
for i in range(0, 7):
    n = int(input())
    if n % 2 == 0:
        s = s + n
print(s)

### 4

На обработку поступает натуральное число. Нужно написать программу, которая выводит на экран максимальную цифру числа, кратную $3$. Если в числе нет цифр, кратных $3$, требуется на экран вывести «NO». Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их ровно $5$). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

**Примечание 1.** Число $0$ делится на любое натуральное число.

**Примечание 2.** При необходимости вы можете добавить нужные строки кода.

In [ ]:
n = int(input())
max_digit = -1
while n > 0:
    digit = n % 10
    if digit % 3 == 0:
        if digit >= max_digit:
            max_digit = digit
    n = n // 10
if max_digit == -1:
    print('NO')
else:
    print(max_digit)

### 5

На обработку поступает натуральное число. Нужно написать программу, которая выводит на экран его первую (старшую) цифру. Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их ровно $2$). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

In [ ]:
n = int(input())
while n > 0:
    digit = n % 10
    n = n // 10
print(digit)

In [ ]:
# решение не по условию, но элегантное
print(input()[0])

### 6

На обработку поступает натуральное число. Нужно написать программу, которая выводит на экран произведение цифр введенного числа. Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их ровно $3$). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

In [ ]:
n = int(input())
product = 1
while n > 0:
    digit = n % 10
    product = product * digit
    n = n // 10
print(product)

## Вложенные циклы. Часть 1

**Примеры**

In [ ]:
for hours in range(24):
    for minutes in range(60):
        for seconds in range(60):
            print(hours, ':', minutes, ':', seconds)

### 1

Дано натуральное число $n$, $(n≤ 9)$. Напишите программу, которая печатает таблицу размером $n×3$ состоящую из данного числа (числа отделены одним пробелом).

In [112]:
n = int(input())
for i in range(n): # rows
    for j in range(3): # columns
        print(n, end=' ')
    print()

4
4 4 4 
4 4 4 
4 4 4 
4 4 4 


### 2

Дано натуральное число $n$, $(n≤ 9)$. Напишите программу, которая печатает таблицу размером $n×5$, где в $i$-ой строке указано число $i$ (числа отделены одним пробелом).

In [119]:
n = int(input())
for i in range(1, n+1):
    for j in range(1, 6): 
        print(i, end=' ')
    print()

5
1 1 1 1 1 
2 2 2 2 2 
3 3 3 3 3 
4 4 4 4 4 
5 5 5 5 5 


### 3

Дано натуральное число $n$, $(n≤ 9)$. Напишите программу, которая печатает таблицу сложения для всех чисел от $1$ до $n$ в соответствии с примером.

In [122]:
n = int(input())
for i in range(1, n+1):
    for j in range(1, 10): 
        print(f"{i} + {j} = {i + j}", end='\n')
    print()

3
1 + 1 = 2
1 + 2 = 3
1 + 3 = 4
1 + 4 = 5
1 + 5 = 6
1 + 6 = 7
1 + 7 = 8
1 + 8 = 9
1 + 9 = 10

2 + 1 = 3
2 + 2 = 4
2 + 3 = 5
2 + 4 = 6
2 + 5 = 7
2 + 6 = 8
2 + 7 = 9
2 + 8 = 10
2 + 9 = 11

3 + 1 = 4
3 + 2 = 5
3 + 3 = 6
3 + 4 = 7
3 + 5 = 8
3 + 6 = 9
3 + 7 = 10
3 + 8 = 11
3 + 9 = 12



### 4

Дано нечетное натуральное число $n$. Напишите программу, которая печатает равнобедренный звездный треугольник с основанием, равным $n$ в соответствии с примером:

In [134]:
n = int(input())
middle = n // 2 + 1 # середина треугольника, после которой начинается убывание звездочек
for i in range(1, n+1):
    if i <= middle:
        print('*' * i)
    else:
        print('*' * (n - i + 1))

7
*
**
***
****
***
**
*


In [135]:
n = int(input())
for i in range(1, n + 1):
    for j in range(i):
        if i + j <= n:
            print('*', end='')
    print()

7
*
**
***
****
***
**
*


### 5

Дано натуральное число $n$. Напишите программу, которая печатает численный треугольник в соответствии с примером:

In [142]:
n = int(input())
for i in range(1, n + 1):
    for j in range(1, i + 1):
        print(i, end='')
    print()

4
1
22
333
4444


**Решение уравнений**

Найдите все пары натуральных чисел (и их количество) являющихся решением уравнения $12x+13y=777$.

Поскольку по условию числа $x$ и $y$ являются натуральными, то $x≤64,y≤59$. Напишем программу, которая перебирает всевозможные пары чисел $(x;y)$ и проверяет для них выполнение условия $12x+13y=777$.

`Если предположить, что x или y равно 0, то 12x+13y=777 примет вид 12x=777 или 13y=777.
Из этого равенства и знания что x и y - натуральные числа, делаем вывод, что делать нужно без остатка или:
x = 777 // 12 = 64 или
y = 777 // 13 = 59 .
Мы предположили, что одно из слагаемых приняло минимальное значение (0), значит значения, которые мы получили - теоретически максимально возможное. Поэтому x <= 64 и y <= 59.`

In [143]:
total = 0
for x in range(1, 65):
    for y in range(1, 60):
        if 12 * x + 13 * y == 777:
            total += 1
            print('x =', x, 'y =', y)
print('Общее количество натуральных решений =', total)

x = 3 y = 57
x = 16 y = 45
x = 29 y = 33
x = 42 y = 21
x = 55 y = 9
Общее количество натуральных решений = 5


Найдите все пары натуральных чисел (и их количество) являющихся решением уравнения $x^2+y^2+z^2 = 2020$.

Поскольку по условию числа $x$, $y$ и $z$ являются натуральными, то $x, y, z \le \sqrt{2020} = 45$. Напишем программу, которая перебирает всевозможные тройки чисел $(x;y;z)$ и проверяет для них условие $x^2+y^2+z^2 = 2020$.

In [144]:
total = 0
for x in range(1, 45):
    for y in range(1, 45):
        for z in range(1, 45):
            if x ** 2 + y ** 2 + z ** 2 == 2020:
                total += 1
                print('x =', x, 'y =', y, 'z =', z)
print('Общее количество натуральных решений =', total)

x = 18 y = 20 z = 36
x = 18 y = 36 z = 20
x = 20 y = 18 z = 36
x = 20 y = 36 z = 18
x = 36 y = 18 z = 20
x = 36 y = 20 z = 18
Общее количество натуральных решений = 6


### 6

Решите уравнение в натуральных числах $28n+30k+31m=365$.

In [149]:
#28n=365 or 30k=365 or 31m=365
n = 365 // 28 #13
k = 365 // 30 #12
m = 365 // 31 #11

total = 0
for n in range(1, 14):
    for k in range(1, 13):
        for m in range(1, 12):
            if 28 * n + 30 * k + 31 * m == 365:
                total += 1
                print('n =', n, 'k =', k, 'm =', m)
print('Общее количество натуральных решений =', total)

n = 1 k = 4 m = 7
n = 2 k = 1 m = 9
Общее количество натуральных решений = 2


### 7

Имеется $100$ рублей. Сколько быков, коров и телят можно купить на все эти деньги, если плата за быка – $10$ рублей, за корову – $5$ рублей, за теленка – $0.5$ рубля и надо купить $100$ голов скота?

In [150]:
# 10*bull + 5*cow + 0.5*calf = 100 (rubles)
# bull + cow + calf = 100 (animals)

total = 0
for bull in range(100):
    for cow in range(100):
        for calf in range(100):
            if 10 * bull + 5 * cow + 0.5 * calf == 100 and bull + cow + calf == 100:
                total += 1
                print('bull =', bull, 'cow =', cow, 'calf =', calf)
print('Общее количество натуральных решений =', total)

bull = 1 cow = 9 calf = 90
Общее количество натуральных решений = 1


### 8

В 1769 году Леонард Эйлер сформулировал обобщенную версию Великой теоремы Ферма, предполагая, что по крайней мере $n$ энных степеней необходимо для получения суммы, которая сама является энной степенью для $n>2$. Напишите программу для опровержения гипотезы Эйлера (продержавшейся до 1967 года), и найдите четыре положительных целых числа, сумма 5-х степеней которых равна 5-й степени другого положительного целого числа.

Таким образом, найдите пять натуральных чисел $a,b,c,d,e$ удовлетворяющих условию:

$a^5+b^5+c^5+d^5=e^5$.

В ответе укажите сумму $a+b+c+d+e$.

**Примечание 1.** Используйте вложенный цикл for.

**Примечание 2.** Считайте, что числа $a, b, c, d, e$ не превосходят $150$.

**Примечание 3.** Программа может работать дольше чем обычно. В зависимости от способа решения задачи на выполнение программы может уходить до нескольких минут. Попробуйте сократить количество вложенных циклов. 

In [1]:
total = 0
while total < 1:
    for a in range(1, 151):
        for b in range(a, 151):
            for c in range(b, 151):
                for d in range(c, 151):
                    for e in range(d, 151):
                        if a**5 + b**5 + c**5 + d**5 == e**5:
                            print('a =', a, 'b =', b, 'c =', c, 'd =', d, 'e =', e)
                            total += 1
                            print('sum =', a + b + c + d + e)

a = 27 b = 84 c = 110 d = 133 e = 144
sum = 498


## Вложенные циклы. Часть 2

### 1

Дано натуральное число $n$. Напишите программу, которая печатает численный треугольник с высотой равной $n$, в соответствии с примером:

In [58]:
n = int(input())
step = 1
for i in range(1, n + 1):
    for j in range(1, i + 1):
        print(step, end=' ')
        step += 1
    print()

5
1 
2 3 
4 5 6 
7 8 9 10 
11 12 13 14 15 


### 2

Дано натуральное число $n$. Напишите программу, которая печатает численный треугольник с высотой равной $n$, в соответствии с примером:

In [77]:
n = int(input())
middle = 1
for i in range(1, n + 1):
    for j in range(1, i + 1):
        print(j, end='')
        middle = i
    while middle > 1:
        print(middle - 1, end='')
        middle -= 1
    print()

5
1
121
12321
1234321
123454321


### 3

На вход программе подается два натуральных числа $a$ и $b$ $(a < b)$. Напишите программу, которая находит натуральное число из отрезка $[a;b]$ с максимальной суммой делителей.

На вход программе подаются два числа, каждое на отдельной строке. Программа должна вывести два числа на одной строке, разделенных пробелом: число с максимальной суммой делителей и сумму его делителей. Если таких чисел несколько, то выведите наибольшее из них.

In [94]:
a, b = int(input()), int(input())
sum = 0
maxsum = 0
solution_number = 0
solution_sum = 0
for number in range(a, b + 1):
    for divider in range(1, number + 1):
        if number % divider == 0:
            sum += divider
    if sum >= maxsum:
        maxsum = sum
        solution_number, solution_sum = number, sum
    sum = 0
print(solution_number, solution_sum)

1
100
96 252


### 4

На вход программе подается натуральное число $n$. Напишите программу, выводящую графическое изображение делимости чисел от $1$ до $n$ включительно. В каждой строке надо напечатать очередное число и столько символов «+», сколько делителей у этого числа.

In [98]:
n = int(input())
counter = 0
for number in range(1, n + 1):
    for divider in range(1, number + 1):
        if number % divider == 0:
            counter += 1
    print(number, '+' * counter, sep='')
    counter = 0

5
1+
2++
3++
4+++
5++


### 5

На вход программе подается натуральное число $n$. Напишите программу, которая находит цифровой корень данного числа. Цифровой корень числа $n$ получается следующим образом: если сложить все цифры этого числа, затем все цифры найденной суммы и повторить этот процесс, то в результате будет получено однозначное число (цифра), которое и называется цифровым корнем данного числа.

In [102]:
n = int(input())
while n > 9:
    sum = 0
    while n > 0:
        last_digit = n % 10
        sum += last_digit
        n = n // 10
    n = sum    
print(n)

192
3


### 6

Дано натуральное число $n$. Напишите программу, которая выводит значение суммы $1!+2!+3!+…+n!$.

In [6]:
n = int(input())
fac = 1
sum = 0
for i in range(1, n + 1):
    for j in range(1, i + 1):
        fac *= j
    sum += fac
    fac = 1
print(sum)

3
9


### 7

На вход программе подается два натуральных числа $a$ и $b$ $(a< b)$. Напишите программу, которая находит все простые числа от $a$ до $b$ включительно.

In [15]:
a, b = int(input()), int(input())
counter = 0
for number in range(a, b + 1):
    for divider in range (1, number + 1):
        if number % divider == 0:
            counter += 1
    if counter == 2:
        print(number)
    counter = 0

1
5
2
3
5


## Итоговая работа на циклы

### 1

На обработку поступает натуральное число. Нужно написать программу, которая выводит на экран сумму чётных цифр этого числа или 0, если чётных цифр в записи нет. Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их может быть одна или несколько). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

In [25]:
n = int(input())
s = 0
while n > 0:
    if n % 2 == 0:
        s += n % 10
    n //= 10
print(s)

22
4


### 2

На обработку поступает последовательность из 8 целых чисел. Известно, что вводимые числа по абсолютной величине не превышают $10^{12}$. Нужно написать программу, которая выводит на экран количество делящихся нацело на 4 чисел в исходной последовательности и максимальное делящееся нацело на 4 число. Если делящихся нацело на 4 чисел нет, требуется на экран вывести «NO». Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их может быть одна или несколько). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

In [ ]:
n = 8
count = 0
maximum = -10**12
for i in range(1, n + 1):
    x = int(input())
    if x % 4 == 0:
        count += 1
        if x > maximum:
            maximum = x
if count > 0:
    print(count)
    print(maximum)
else:
    print('NO')

### 3

На обработку поступает последовательность из 4 целых чисел. Известно, что вводимые числа по абсолютной величине не превышают $10^8$. Нужно написать программу, которая выводит на экран количество нечётных чисел в исходной последовательности и максимальное нечётное число. Если нечётных чисел нет, требуется на экран вывести «NO». Программист торопился и написал программу неправильно.

Найдите все ошибки в этой программе (их может быть одна или несколько). Известно, что каждая ошибка затрагивает только одну строку и может быть исправлена без изменения других строк.

In [ ]:
n = 4
count = 0
maximum = -10**8
for i in range(1, n + 1):
    x = int(input())
    if x % 2 != 0:
        count += 1
        if x > maximum:
            maximum = x
if count > 0:
    print(count)
    print(maximum)
else:
    print('NO')

### 4

На вход программе подается натуральное число $n$. Напишите программу, которая печатает звездную рамку размерами $n×19$. На вход программе подаётся натуральное число $n∈[3;19]$ — высота звездной рамки.

In [31]:
n = int(input())
for star in range(1, n + 1):
    if star != 1 and star != n:
        print('*' + ' ' * 17 + '*')
    else:
        print('*' * 19)

4
*******************
*                 *
*                 *
*******************


### 5

Дано натуральное число $n$ $(n>99)$. Напишите программу, которая определяет его третью (с начала) цифру.

In [35]:
n = int(input())
while n > 99:
    last_digit = n % 10
    n = n // 10
print(last_digit)

15230
2


### 6

Дано натуральное число. Напишите программу, которая вычисляет:

* количество цифр 3 в нем;
* сколько раз в нем встречается последняя цифра;
* количество четных цифр;
* сумму его цифр, больших пяти;
* произведение цифр, больших семи (если цифр больших семи нет, то вывести 1, если такая цифра одна, то вывести ее);
* сколько раз в нем встречается цифры 0 и 5 (всего суммарно).

In [26]:
n = int(input())
count_3, count_lastdigit, count_even, sum_more5, mult_more7, count_05 = 0, 0, 0, 0, 1, 0
last = n % 10
while n != 0:
    last_digit = n % 10  
    if last_digit == 3:
        count_3 += 1
    if last_digit == last:
        count_lastdigit += 1
    if last_digit % 2 == 0:
        count_even += 1
    if last_digit > 5:
        sum_more5 += last_digit
    if last_digit > 7:
        mult_more7 *= last_digit
    if last_digit == 0 or last_digit == 5:
        count_05 += 1
    n = n // 10 
print(count_3, count_lastdigit, count_even, sum_more5, mult_more7, count_05, sep='\n')

56639
1
1
2
21
9
1


### 7

Сриниваса Рамануджан – индийский математик, славившийся своей интуицией в области чисел. Когда английский математик Годфри Харди навестил его однажды в больнице, он обмолвился, что номером такси, на котором он приехал, было $1729$, такое скучное и заурядное число. На что Рамануджан ответил: "Нет, нет! Это очень интересное число. Это наименьшее число, выражаемое как сумма двух кубов двумя разными способами". Другими словами:

$1729 = 1^3 + 12^3 = 9^3 + 10^3$.

Напишите программу, которая находит аналогичные интересные числа. В ответе запишите первые 5 чисел в порядке возрастания, включая число $1729$.

**Примечание.** Используйте вложенный цикл.

In [ ]:
for i in range(1, 40):
    for j in range(1, 40):
        for k in range(1, 40):
            for l in range(1, 40):
                if i != k and i != l and j != k and j != l and i ** 3 + j ** 3 == k ** 3 + l ** 3:
                    print(i ** 3 + j ** 3)